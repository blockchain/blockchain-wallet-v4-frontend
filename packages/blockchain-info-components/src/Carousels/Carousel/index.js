import React from 'react'
import PropTypes from 'prop-types'
import CarouselTemplate from './template.js'

class Carousel extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = { index: 0, total: props.children.length - 1 }
    this.handleClick = this.handleClick.bind(this)
    this.handlePrevious = this.handlePrevious.bind(this)
    this.handleNext = this.handleNext.bind(this)
    this.interval = undefined
  }

  componentDidMount () {
    const { auto, delay } = this.props
    if (auto) {
      this.interval = setInterval(
        () => this.setState({ index: this.getNextIndex() }),
        delay
      )
    }
  }

  componentWillUnmount () {
    if (this.interval) {
      clearInterval(this.interval)
    }
  }

  getPreviousIndex () {
    const { index, total } = this.state
    return index === 0 ? total : index - 1
  }

  getNextIndex () {
    const { index, total } = this.state
    return index === total ? 0 : index + 1
  }

  handleClick (index) {
    this.setState({ index })
  }

  handlePrevious () {
    this.setState({ index: this.getPreviousIndex() })
  }

  handleNext () {
    this.setState({ index: this.getNextIndex() })
  }

  render () {
    const { children, ...rest } = this.props

    return (
      <CarouselTemplate
        index={this.state.index}
        total={this.state.total}
        handleClick={this.handleClick}
        handlePrevious={this.handlePrevious}
        handleNext={this.handleNext}
        {...rest}
      >
        {children}
      </CarouselTemplate>
    )
  }
}

Carousel.propTypes = {
  auto: PropTypes.bool,
  delay: PropTypes.number,
  height: PropTypes.number,
  arrows: PropTypes.bool,
  chips: PropTypes.bool
}

Carousel.defaultProps = {
  auto: false,
  delay: 3000,
  height: 150,
  arrows: true,
  chips: true
}

export default Carousel
