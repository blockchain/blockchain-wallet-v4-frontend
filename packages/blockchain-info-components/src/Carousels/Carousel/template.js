import React from 'react'
import { FormattedMessage } from 'react-intl'
import PropTypes from 'prop-types'
import { range } from 'ramda'
import styled from 'styled-components'

import { Button } from '../../Buttons'
import { Color } from '../../Colors/index.ts'
import { Icon } from '../../Icons'
import { Text } from '../../Text'

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
`
const Content = styled.div`
  position: relative;
  width: 100%;
  min-height: ${props => props.height}px;
  overflow-x: hidden;
`
const Slider = styled.div`
  position: absolute;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  top: 0;
  left: ${props => props.index * -100}%;
  width: ${props => (props.total + 1) * 100}%;
  min-height: auto;
  transition: all 0.4s ease-in-out;

  & > * {
    position: relative;
    width: 100%;
    height: 100%;
  }
`
const Controls = styled.div`
  position: relative;
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 40px;
`
const Control = styled.div`
  display: flex;
  width: 10px;
  height: 10px;
  margin: 0 10px;
  border-radius: 5px;
  background-color: ${props =>
    props.active ? props.theme.blue600 : props.theme['grey500']};
  cursor: pointer;
  z-index: ;
`
const Arrow = styled(Icon).attrs({
  name: 'chevron-down',
  size: '34px',
  weight: 700,
  color: 'grey800'
})`
  position: absolute;
  top: 50%;
  margin-top: -18px;
  left: ${props => (props.left ? '10px' : 'initial')};
  right: ${props => (!props.left ? '10px' : 'initial')};
  transform: ${props => (props.left ? 'rotate(90deg)' : 'rotate(-90deg)')};
  cursor: pointer;
  z-index: 1;

  &:hover {
    color: ${Color('blue600')}!important;
  }
`
const FooterWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  & > :first-child {
    margin-right: 10px;
  }
`

const Carousel = props => {
  const {
    arrows,
    children,
    chips,
    handleClick,
    handleNext,
    handlePrevious,
    height,
    index,
    nextButton,
    total
  } = props

  return (
    <Wrapper>
      <Content height={height} total={total}>
        {arrows && <Arrow left onClick={handlePrevious} />}
        <Slider height={height} index={index} total={total}>
          {children}
        </Slider>
        {arrows && <Arrow onClick={handleNext} />}
      </Content>
      {chips && (
        <Controls>
          {range(0, total + 1).map((value, arrayIndex) => (
            <Control
              key={arrayIndex}
              onClick={() => handleClick(value)}
              active={index === value}
            />
          ))}
        </Controls>
      )}
      {nextButton && (
        <FooterWrapper>
          {index < total && (
            <Button width='60px' nature='light' onClick={handleNext}>
              <FormattedMessage
                id='components.emptyscenecarousel.next'
                defaultMessage='Next'
              />
            </Button>
          )}
          {index === total && (
            <Button width='70px' nature='light' onClick={handleNext}>
              <FormattedMessage id='buttons.go_back' defaultMessage='Go Back' />
            </Button>
          )}
          <Text size='14px' weight={600} color='grey400'>
            {index + 1} of {total + 1}
          </Text>
        </FooterWrapper>
      )}
    </Wrapper>
  )
}

Carousel.propTypes = {
  index: PropTypes.number.isRequired,
  total: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  handleClick: PropTypes.func.isRequired,
  arrows: PropTypes.bool.isRequired,
  chips: PropTypes.bool.isRequired
}

export default Carousel
