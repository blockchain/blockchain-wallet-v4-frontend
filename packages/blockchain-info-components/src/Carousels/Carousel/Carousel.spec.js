import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Carousel from './template'

describe('Carousel component', () => {
  it('default renders correctly', () => {
    const component = shallow(
      <Carousel
        height={12}
        chips={false}
        total={12}
        index={4}
        arrows
        handleClick={() => {}}
      />
    )
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })
})
