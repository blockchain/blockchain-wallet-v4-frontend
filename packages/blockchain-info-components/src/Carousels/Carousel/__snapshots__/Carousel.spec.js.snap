// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel component default renders correctly 1`] = `
<template__Wrapper>
  <template__Content
    height={12}
    total={12}
  >
    <template__Arrow
      left={true}
    />
    <template__Slider
      height={12}
      index={4}
      total={12}
    />
    <template__Arrow />
  </template__Content>
</template__Wrapper>
`;
